File: src/main/java/no/novari/flyt/configuration/validation/instancereference/InstanceValueTypeCompatibilityValidator.java
package no.novari.flyt.configuration.validation.instancereference;

import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto;
import no.novari.flyt.configuration.model.configuration.entities.ValueMapping;
import no.novari.flyt.configuration.model.metadata.InstanceValueMetadata;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;

public interface InstanceValueTypeCompatibilityValidator {

    default List<Tuple2<String, InstanceValueMetadata.Type>> findIncompatibleInstanceValuesKeyAndType(
            ValueMappingDto valueMappingDto,
            Collection<String> referencedInstanceValueKeys,
            Map<String, InstanceValueMetadata.Type> instanceValueTypePerKey
    ) {
        if (valueMappingDto.getType() != getTypeToValidate()) {
            return Collections.emptyList();
        }
        return referencedInstanceValueKeys.stream().filter(key -> !getCompatibleTypes().contains(instanceValueTypePerKey.get(key))).map(key -> Tuples.of(key, instanceValueTypePerKey.get(key))).toList();
    }

    ValueMapping.Type getTypeToValidate();

    Set<InstanceValueMetadata.Type> getCompatibleTypes();

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/instancereference/InstanceValueTypeCompatibilityValidator.kt
package no.novari.flyt.configuration.validation.instancereference

import no.novari.flyt.configuration.model.configuration.dtos.ValueMappingDto
import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import no.novari.flyt.configuration.model.metadata.InstanceValueMetadata
import java.util.AbstractMap

interface InstanceValueTypeCompatibilityValidator {
    fun findIncompatibleInstanceValuesKeyAndType(
        valueMappingDto: ValueMappingDto,
        referencedInstanceValueKeys: Collection<String>,
        instanceValueTypePerKey: Map<String, InstanceValueMetadata.Type>,
    ): List<Map.Entry<String, InstanceValueMetadata.Type>> {
        if (valueMappingDto.type != getTypeToValidate()) {
            return emptyList()
        }

        return referencedInstanceValueKeys
            .filter { key -> !getCompatibleTypes().contains(instanceValueTypePerKey[key]) }
            .mapNotNull { key -> instanceValueTypePerKey[key]?.let { type -> AbstractMap.SimpleEntry(key, type) } }
    }

    fun getTypeToValidate(): ValueMapping.Type

    fun getCompatibleTypes(): Set<InstanceValueMetadata.Type>
}
